src/components/CreateAccount.js

```diff
@@ -56,59 +71,230 @@ class CreateAccount extends Component {
+            });
+            break;
+        case 'email':
+            this.setState({
+                'emailIsValid': this.validateEmail(value)
+            });
+            break;
+        case 'password':
+            this.setState({
+                'passwordIsValid': this.validatePassword(value)
+            });
+            break;
+        }
+
         this.setState({
-            [event.target.name]: event.target.value
+            [event.target.name]: value,
         });
     }
 
+    /**
+     * Returns true if the username is valid and false otherwise
+     * @param {String} username
+     */
+    validateUsername(username) {
+        if (username.length < 3 || username.length > 20) {
+            return false;
+        }
+        const regexes = [
+            /[A-Z]+/,
+            /[a-z]+/,
+            /[0-9]+/,
+            /[.\-_]+/
+        ];
+        for (let i in regexes) {
+            let match = username.match(regexes[i]);
+            while (match !== null) {
+                username = username.replace(match[0], '');
+                match = username.match(regexes[i]);
+            }
+        }
+        return username.length === 0;
+    }
+
+    /**
+     * Returns true if the password is valid and false otherwise
+     * @param {String} password
+     */
+    validatePassword(password) {
+        if (password.length < 8) {
+            return false;
+        }
+        const regexes = [
+            /[A-Z]+/,
+            /[a-z]+/,
+            /[0-9]+/,
+            /[.\-!@#$%^&*?_+ ]+/
+        ];
+        for (let i in regexes) {
+            let match = password.match(regexes[i]);
+            let count = 0;
+            while (match !== null) {
+                password = password.replace(match[0], '');
+                match = password.match(regexes[i]);
+                count++;
+            }
+            if (count === 0) {
+                return false;
+            }
+        }
+        return password.length === 0;
+    }
+
+    /**
+     * Returns true if the email is valid and false otherwise
+     * @param {String} email
+     */
+    validateEmail(email) {
+        const re = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
+        return re.test(email.toLowerCase());
+    }
+
+    /**
+     * Triggers a redirect to the SignIn page when the user hits the
+     * "Sign In" button.
+     * @param {Event} event
+     */
+    handleSignInClick(event) {
+        event.preventDefault();
+        this.setState({ goToSignIn: true });
+    }
+
     render() {
-        let errorMsg;
+        if (this.state.isSignedIn) {
+            return <Redirect to='/' />;
+        } else if (this.state.goToSignIn) {
+            return <Redirect to='/sign-in' />;
+        }
+
+        let errorMsg, usernameWarning, emailWarning, passwordWarning;
+        let buttonClass = 'btn btn-primary';
```

Comment:
put this outside of render() otherwise React will re-process these variables again everytime the component re-renders due to state / prop changes